File: python/euler/bin/p0062.py
# project euler: problem 62

from itertools import count


def compute(n_of_perms: int) -> str:
    def make_key(n: int) -> str:
        return ''.join(sorted(list(str(n))))

    tbl: dict[str, list[int]] = dict()
    for n in count(1):
        cube = n * n * n
        if (key := make_key(cube)) in tbl:
            tbl[key] = tbl[key] + [n]
            if len(tbl[key]) == n_of_perms:
                return str(tbl[key][0] ** 3)
        else:
            tbl[key] = [n]

    raise RuntimeError('unreachable!')


def solve() -> str:
    return compute(5)

File: typescript/euler/bin/p0062.ts
// project euler: problem 62

export const compute = (numOfPerms: number): string => {
  const makeKey = (n: number): string => String(n).split("").sort().join("");

  const tbl = new Map<string, number[]>();
  let n = 0;
  while (true) {
    n += 1;
    const key = makeKey(n * n * n);
    if (tbl.has(key)) {
      tbl.set(key, tbl.get(key)!.concat(n));
      if (tbl.get(key)!.length === numOfPerms) {
        return String(tbl.get(key)![0] ** 3);
      }
    } else {
      tbl.set(key, [n]);
    }
  }
};

export const solve = (): string => compute(5);
